sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/BeamFnDataGrpcClient.java

```diff
@@ -0,0 +1,123 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link BeamFnDataClient} that uses gRPC for sending and receiving data.
+ *
+ * <p>TODO: Handle closing clients that are currently not a consumer nor are being consumed.
+ */
+public class BeamFnDataGrpcClient implements BeamFnDataClient {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BeamFnDataGrpcClient.class);
+
+  private final ConcurrentMap<BeamFnApi.ApiServiceDescriptor, BeamFnDataGrpcMultiplexer> cache;
+  private final Function<BeamFnApi.ApiServiceDescriptor, ManagedChannel> channelFactory;
+  private final BiFunction<Function<StreamObserver<BeamFnApi.Elements>,
+                                    StreamObserver<BeamFnApi.Elements>>,
+                           StreamObserver<BeamFnApi.Elements>,
+                           StreamObserver<BeamFnApi.Elements>> streamObserverFactory;
+  private final PipelineOptions options;
+
+  public BeamFnDataGrpcClient(
+      PipelineOptions options,
+      Function<BeamFnApi.ApiServiceDescriptor, ManagedChannel> channelFactory,
+      BiFunction<Function<StreamObserver<BeamFnApi.Elements>, StreamObserver<BeamFnApi.Elements>>,
+                 StreamObserver<BeamFnApi.Elements>,
+                 StreamObserver<BeamFnApi.Elements>> streamObserverFactory) {
+    this.options = options;
+    this.channelFactory = channelFactory;
+    this.streamObserverFactory = streamObserverFactory;
+    this.cache = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * Registers the following inbound stream consumer for the provided instruction id and target.
+   *
+   * <p>The provided coder is used to decode elements on the inbound stream. The decoded elements
+   * are passed to the provided consumer. Any failure during decoding or processing of the element
+   * will complete the returned future exceptionally. On successful termination of the stream
+   * (signaled by an empty data block), the returned future is completed successfully.
+   */
+  @Override
+  public <T> CompletableFuture<Void> forInboundConsumer(
+      BeamFnApi.ApiServiceDescriptor apiServiceDescriptor,
+      KV<Long, BeamFnApi.Target> inputLocation,
+      Coder<WindowedValue<T>> coder,
+      ThrowingConsumer<WindowedValue<T>> consumer) {
+    LOGGER.debug("Registering consumer instruction {} for target {}",
+        inputLocation.getKey(),
+        inputLocation.getValue());
+
+    CompletableFuture<Void> readFuture = new CompletableFuture<>();
+    BeamFnDataGrpcMultiplexer client = getClientFor(apiServiceDescriptor);
+    client.futureForKey(inputLocation).complete(
+        new BeamFnDataInboundObserver<>(coder, consumer, readFuture));
+    return readFuture;
+  }
+
+  /**
+   * Creates a closeable consumer using the provided instruction id and target.
+   *
+   * <p>The provided coder is used to encode elements on the outbound stream. The encoded elements
+   * are buffered upto the provided buffer limit before being emit on the outbound stream.
```

Comment:
Fixed, older implementation used to allow sending the buffer limit along this method, now we use an explicit outbound observer responsible for buffering.